src/community/wps-download/src/main/java/org/geoserver/wps/gs/download/RasterDownload.java

```diff
@@ -178,170 +185,441 @@ public Resource execute(String mimeType, final ProgressListener progressListener
+            for (GeneralParameterValue readParameter : readParameters) {
+                if ("BackgroundValues"
+                        .equalsIgnoreCase(readParameter.getDescriptor().getName().toString())) {
+                    Object bgValue = ((ParameterValue) readParameter).getValue();
+                    if (bgValue != null && bgValue instanceof double[]) {
+                        backgroundValues = ((double[]) bgValue);
+                    }
+                    break;
+                }
             }
         }
+        return backgroundValues;
+    }
+
+    /**
+     * Check if a crop operation is needed (meaning that the extent of the returned image contains 
+     * the requested extent).
+     * 
+     * @param rasterData
+     * @param requestedRange
+     * @return
+     */
+    private boolean cropIsNeeded(RenderedImage rasterData, GridEnvelope requestedRange) {
+        final int requestedW = requestedRange.getSpan(0);
+        final int requestedH = requestedRange.getSpan(1);
+        final int requestedMinX = requestedRange.getLow(0);
+        final int requestedMinY = requestedRange.getLow(1);
+        final int requestedMaxX = requestedRange.getHigh(0);
+        final int requestedMaxY = requestedRange.getHigh(1);
+
+        final int imageW = rasterData.getWidth();
+        final int imageH = rasterData.getHeight();
+        final int minX = rasterData.getMinX();
+        final int minY = rasterData.getMinY();
+        final int maxX = minX + imageW - 1;
+        final int maxY = minY + imageH - 1;
+
+        if (((imageW + ((minX >= 0) ? 0 : minX)) >= requestedW)
+                && ((imageH + ((minY >= 0) ? 0 : minY)) >= requestedH)
+                && ((requestedMinX >= minX) && (requestedMinY >= minY) && (requestedMaxX <= maxX)
+                        && (requestedMaxY <= maxY))) {
+            // The extent of the returned image contains the requested extent.
+            // Let's do a crop
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * The requested region is wider than the returned coverage (the reader may have done a crop to minimal extent).
+     * Extend the result to the requested region, using background value.
+     * 
+     * @param rasterData
+     * @param requestedGridGeometry
+     * @param backgroundValues
+     * @return
+     */
+    private GridCoverage2D extendToRegion(RenderedImage rasterData,
+            GridGeometry2D requestedGridGeometry, double[] backgroundValues) {
+
+        final GridEnvelope requestedRange = (GridEnvelope) requestedGridGeometry.getGridRange();
+        final int requestedW = requestedRange.getSpan(0);
+        final int requestedH = requestedRange.getSpan(1);
+
+        ImageLayout layout = new ImageLayout();
+        // Setting output image layout
+        // I can impose 0,0 as the origin since we are working under
+        // that assumption as part of the reader request contract
```

Comment:
reader request contract? Afaik there is no such a thing, believe I've modified mosaic to avoid super-large initial offsets just recently to avoid math issues.

Reply:
Fixed